### ISSUE TITLE
NMS implementation does not scale

### ISSUE THREAD

**Issue by GFleishman** [mpicbg-csbd/stardist#39]
I've proposed a solution to this issue in a pull request (pending).

The [NMS implementation](https://github.com/mpicbg-csbd/stardist/blob/f8e63dc3fe1c662b4ee71b2cc54d01728b8bd6a5/stardist/lib/stardist3d_impl.cpp#L929-L1306) scales with the square of the number of candidate polygons, i.e. for N candidate polygons the average case is still O(N**2). This is acceptable in 2D because comparing two candidate polygons for overlap, even if they must be rendered completely, is fast. This is also acceptable in 3D _for sufficiently small N_ since the inner loop of the algorithm is parallelized over threads and some smart lower/upper bounds are used to potentially identify a suppression without rendering the full polygons.

However, some potentially common and impactful use cases do not have sufficiently small N. The 6 days post fertilization larval zebrafish brain, a common laboratory organism, has on the order of 100,000 neurons. See below an example network which, with a probability threshold optimized for a validation dataset, produced 6.32e6 polygon candidates. Making matters worse, at typical confocal or light sheet microscope sampling rates, individual nuclei in whole brain images of this organism comprise relatively few voxels. In that case, to ensure sufficiently accurate boundary reconstruction, [the step size used for polygon rendering must be reduced](https://github.com/mpicbg-csbd/stardist/issues/33#issuecomment-583421466), slowing down reconstruction time for each instance. Instance segmentation in that case took 32 cores 8 hours to complete, 7.25 of which are from the NMS. On our cluster system, this works out to about $20 of chargeback cost.

However, for instance segmentation over such a large field of view, the overwhelming majority of polygon candidate comparisons cannot possibly result in a suppression, as the candidates do not overlap at all - ideally such comparisons should be ignored. In my pull request I propose a solution which produces the same segmentation, but prevents these unproductive candidate comparisons and reduces the total execution time 40 minutes, only 21 of which are from the NMS.

See lines:
NMS: n_polys = 6315439
NMS: using OpenMP with 64 thread(s)
NMS took 25817.3547 s
![full_image_old](https://user-images.githubusercontent.com/8507206/76967275-93912900-68fd-11ea-9f2a-64afde69335b.png)

**Comment by maweigert** [comment]
Hi Greg, 

> In my pull request I propose a solution which produces the same segmentation, but prevents these unproductive candidate comparisons and reduces the total execution time 40 minutes, only 21 of which are from the NMS.

Wow, that would be indeed terrific! How did you do this? Looking forward to the PR very much :))

**Comment by GFleishman** [comment]
@maweigert Just finished the PR :) !

**Comment by GFleishman** [comment]
Oh I also just realized, the approximate times I give for the `cKDTree.query_ball_tree` query times also include passing the list of lists to the c code and converting it to a vector of vectors. My C++ is really rusty, so I might be doing something really inefficient in this conversion (`lists_to_vectors` function in `stardist3d.cpp`), e.g. the push_back function seems suspect to me because it reallocates RAM every time the vector grows. There is likely a better way to write this.

**Comment by constantinpape** [comment]
Just saw the PR #40 and this issue and had to questions:
- Is this still something you are working on? I am asking because the PR is closed.
- Is this something that could also be done in 2d to provide a speed-up? Processing large batches of images still takes a long time with stardist and is cpu bound, I guess because the NMS implementation is relatively slow.

**Comment by uschmidt83** [comment]
> Processing large batches of images still takes a long time with stardist and is cpu bound, I guess because the NMS implementation is relatively slow.

Can you give me some numbers to put this statement in perspective?

We typically don't have issues with the 2D NMS code. It's optimized for cases where the largest (predicted) object size is much smaller than the image size.

**Comment by constantinpape** [comment]
> Can you give me some numbers to put this statement in perspective?

To give you some context:
We run stardist prediction on a DAPI channel to segment nuclei for a high throughput screening setup.
The individual images are approximately 1000x1000 and we have approximately a 1000 images per experiment (= 1 plate). 

Running prediction / segmentation with stardist takes ~ 40 minutes for 1 experiment and 
it's clearly cpu bound (judging from the GPU utilization which I think is maybe between 5 and 10 percent on average). (I just call `predict_instances` for now, so I don't really know what the ratio between time spent prediction and time spent on NMS is).

We also run two other processing steps: predicting foreground/background and boundaries with a pytorch network (on a different channel) and watersheds (using nuclei as seeds and boundary predictions as height map). Both of these steps can be done in under 5 - 10 minutes each, so stardist is the clear bottleneck right now. (For the pytorch prediction I stack images across the batch axis and the watershed can just be parallelized over multiple CPUs, so this comparison is not really fair!)

I can run some experiments next week to really see how long prediction vs NMS takes.
What are the functions I have to call on the model if I want to do this in two separate steps?

**Comment by maweigert** [comment]
Thanks for the context 

1.)
> Running prediction / segmentation with stardist takes ~ 40 minutes for 1 experiment and

How many instances per image do you get approximately? I just did a test a 1024x1024 DSB image with ~1400 instances, and `predict_instances` took 0.7s per image on a single CPU core (i.e. ~12 min for 1000 images). So this is surprising to see.

2.)
> What are the functions I have to call on the model if I want to do this in two separate steps?

Assuming `img` is your (normalized) 2D input, than you would have to call the following two functions: 
```
prob, dist = model.predict(img)  # Step 1
labels, _  = model._instances_from_prediction(img.shape,prob,dist)   # Step 2 (NMS)
```

3.) For many small images (as in your case) it my be faster to parallelise the NMS step explicitly like so (assuming `imgs` is your list of images):

```
from concurrent.futures import ThreadPoolExecutor, ProcessPoolExecutor

def _get_label(shape_prob_dist):
    shape, prob, dist = shape_prob_dist
    return model._instances_from_prediction(shape,prob,dist)[0]
        
shape_prob_dist = tuple((img.shape,)+model.predict(img) for img in imgs)
   
with ProcessPoolExecutor(max_workers=20) as executor:
    labels = tuple(executor.map(_get_label, shape_prob_dist))
```
Its advisable to switch off OpenMP by setting OMP_NUM_THREADS=1.
In my test case, the per image processing time then goes down from 0.7s to 0.27s (so ~4 minutes for 1000images)

Hope that helps,
M

**Comment by GFleishman** [comment]
Chiming in late - and perhaps with less useful information - and I haven't looked at the 2D NMS code:

KDTree could theoretically help in 2D as well, but because rendering the polygons (or whatever other upper/lower bound tricks are present for 2D) is much faster, it may not be necessary and I think Martin's solutions of more compute horsepower is probably just better. KDTree would really only be effective if your images have _many_ cells distributed over a large field of view relative to the size of each individual cell.

Something I've noticed is that networks which are undertrained don't really cluster the probability very tightly to near the centroid of the cell - so you can end up with _many_  polygon candidates per cell. This leads to (relatively) much longer NMS times since there are more polygon candidates to go through.

One difference between your situation and Martin's test case _could_ be that your network needs more training time --> tighter probabilities --> fewer polygon candidates --> shorter NMS time.

As for the closed PR: there was a lot of productive back and forth between Martin and I about how best to get KDTree and a few other efficiency features available for the 3D library and it's still WIP. I stepped back for the past week to run some experiments on my own data, but I'm still willing to help get it all written.

**Comment by constantinpape** [comment]
@maweigert 
Thanks for your last post, I have implemented this now and it speeds up the stardist prediction significantly. I don't have exact timings yet; I am currently implementing proper logging for our workflow and will get this afterwards. Will let you know once I have it.

@GFleishman 
Thanks for clarifying! For our problem, the solution Martin proposed is def. the way to go.
But might be interesting to keep the KDTree in mind for applying stardist to large 2d images, e.g. from a whole slide scanner.

**Comment by nalinratnayeke** [comment]
@maweigert Just revisiting this after finally getting a chance to sit down and try to incorporate Stardist into our workflow. We've been impressed by how well Stardist works on our data (even with the pretrained model), so we're trying tie in Stardist for our high-throughput fixed and timelapse imaging experiments.

I tried out the parallelization of the NMS step and ran into some slight issues getting it to run correctly. It seems like each worker would reload the model and since Tensorflow by default reserves GPU memory it would allocate all the GPU memory and start eating up RAM  (I'm not sure if this is an issue caused by the Tensorflow 2 version of Stardist). I ended up getting things to work alright by allowing GPU growth with csbdeep.utils.tf.limit_gpu_memory() so that even though the model is reloaded (and I assume a new TF session is started?) it doesn't consume all of the memory. Is there a better workaround for this that doesn't involve the model being reloaded when using ProcessPoolExecutor? I suppose making a function that isn't attached to the model class maybe? Attaching some sample code for how I've implemented this, and you could remove the limit_gpu_memory() calls to see the memory leak.
[parallel.zip](https://github.com/mpicbg-csbd/stardist/files/5240536/parallel.zip)

Once we got the parallelization to work, we were able to get the instance calculation down to ~0.4 sec/image, with the network prediction also taking ~0. 4 sec/image for 2000 x 2000 images with a few thousand instances. For our single timepoint fixed imaging, this is not bad since we are usually acquiring ~500-2000 images per experiment, but for our live cell imaging the number of images increases quite a bit (~100 - 300 timepoints X 500 images) so ideally we would want to get the total predict instance time down (we also have some additional processing for tracking that will take time as well).

Not sure what your thoughts on other potential areas of speedup could be, but I thought allowing for batch processing by the prediction step could allow for the GPU to be utilized better. Right now my GPU (RTX Titan) is ~40% utilized during a loop through of predicting images, so I suppose there could be up to a ~2 fold increase in processing speed through strategies of increasing GPU utilization?

I was also thinking about asynchronous processing so that the CPU bound instance processing and GPU bound prediction could happen simultaneously, again affording ~2 fold increase in processing speed. I'm new to python so I don't really know where to start for that, but I was wondering if that was an option you have considered already.

In the end if we aren't able to get any speedups beyond what the NMS parallelization, we might just use Stardist for our fixed imaging, and use a method which doesn't use NMS for determining nuclear instances (e.g. looking for probability peaks for watershed seeds) for timelapse. Since we can use tracking information to correct some missegmentations (i.e. merges and splits), a lower accuracy instance prediction could still be fine for our purposes. Having high fidelity nuclear instances would still be great to have though and would increase our tracking accuracy, so it would be great to use Stardist for our timelapse experiments. In any case, great work with this tool, and I'm looking forward to working with it some more!
